File: backend/src/main/java/com/contentplatform/backend/application/service/TenantRouteService.java
package com.contentplatform.backend.application.service;

import com.contentplatform.backend.application.exception.BadRequestException;
import com.contentplatform.backend.infrastructure.jpa.entity.TenantRouteEntity;
import com.contentplatform.backend.infrastructure.jpa.repository.ApplicationJpaRepository;
import com.contentplatform.backend.infrastructure.jpa.repository.TenantRouteJpaRepository;
import com.contentplatform.backend.domain.value.TenantRouteStatus;
import com.contentplatform.backend.interfaces.web.request.TenantRouteDefinitionRequest;
import com.contentplatform.backend.interfaces.web.request.TenantRouteSyncRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.*;

@Service
public class TenantRouteService {
    private final TenantRouteJpaRepository routeRepo;
    private final ApplicationJpaRepository applicationRepo;
    public TenantRouteService(TenantRouteJpaRepository routeRepo, ApplicationJpaRepository applicationRepo) {
        this.routeRepo = routeRepo; this.applicationRepo = applicationRepo;
    }

    @Transactional
    public Map<String, Object> sync(String applicationId, TenantRouteSyncRequest request) {
        if (!applicationRepo.existsById(applicationId)) throw new BadRequestException("Application not found.");
        String source = required(request.getSource(), "source is required");
        Instant now = Instant.now();
        List<TenantRouteEntity> existing = routeRepo.findByApplicationIdAndSource(applicationId, source);
        Map<String, TenantRouteEntity> byKey = new HashMap<>();
        existing.forEach(route -> byKey.put(route.getRouteKey(), route));
        Set<String> keys = new HashSet<>();
        for (TenantRouteDefinitionRequest definition : request.getRoutes()) {
            String key = required(definition.getKey(), "route key is required");
            String path = required(definition.getPath(), "route path is required");
            if (!keys.add(key)) throw new BadRequestException("Route keys must be unique.");
            if (!path.startsWith("/")) throw new BadRequestException("Route path must start with '/'.");
            Map<String, String> titles = new HashMap<>();
            definition.getTitles().forEach((locale, title) -> {
                if (locale != null && title != null && !locale.isBlank() && !title.isBlank()) titles.put(locale.trim(), title.trim());
            });
            if (titles.isEmpty()) throw new BadRequestException("At least one localized title is required.");
            TenantRouteEntity route = byKey.get(key);
            if (route == null) {
                route = new TenantRouteEntity(UUID.randomUUID().toString(), applicationId, source, key, path, titles,
                    TenantRouteStatus.AVAILABLE, trim(definition.getIcon()), trim(definition.getCssClass()),
                    definition.getMetadata(), now, now, now);
            } else {
                route.synchronize(path, titles, trim(definition.getIcon()), trim(definition.getCssClass()), definition.getMetadata(), now);
            }
            routeRepo.save(route);
        }
        int unavailable = 0;
        if (!Boolean.FALSE.equals(request.getReplaceMissing())) {
            for (TenantRouteEntity route : existing) {
                if (!keys.contains(route.getRouteKey())) { route.markUnavailable(now); routeRepo.save(route); unavailable++; }
            }
        }
        return Map.of("applicationId", applicationId, "source", source, "synchronized", keys.size(),
            "unavailable", unavailable, "synchronizedAt", now.toString());
    }

    public List<TenantRouteEntity> list(String applicationId) {
        if (!applicationRepo.existsById(applicationId)) throw new BadRequestException("Application not found.");
        return routeRepo.findByApplicationIdAndStatusOrderBySourceAscRouteKeyAsc(applicationId, TenantRouteStatus.AVAILABLE);
    }
    private String required(String value, String message) { String result = trim(value); if (result == null) throw new BadRequestException(message); return result; }
    private String trim(String value) { if (value == null || value.trim().isEmpty()) return null; return value.trim(); }
}

File: backend-nestjs/src/services/tenant-route.service.ts
import { BadRequestException, Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { v4 as uuidv4 } from 'uuid';
import { TenantRouteSyncRequestDto } from '../dto/requests/tenant-route-sync-request.dto';
import { ApplicationEntity } from '../entities/application.entity';
import { TenantRouteEntity, TenantRouteStatus } from '../entities/tenant-route.entity';

@Injectable()
export class TenantRouteService {
  constructor(
    @InjectRepository(TenantRouteEntity)
    private readonly routeRepo: Repository<TenantRouteEntity>,
    @InjectRepository(ApplicationEntity)
    private readonly applicationRepo: Repository<ApplicationEntity>,
  ) {}

  async syncByApplicationId(applicationId: string, request: TenantRouteSyncRequestDto) {
    const application = await this.applicationRepo.findOne({ where: { id: applicationId } });
    if (!application) {
      throw new BadRequestException('Application not found.');
    }
    return this.sync(application, request);
  }

  async listByApplicationId(applicationId: string): Promise<TenantRouteEntity[]> {
    return this.routeRepo.find({
      where: { applicationId },
      order: { source: 'ASC', routeKey: 'ASC' },
    });
  }

  async sync(application: ApplicationEntity, request: TenantRouteSyncRequestDto) {
    const source = request.source.trim();
    const keys = new Set<string>();
    const now = new Date();
    const existing = await this.routeRepo.find({ where: { applicationId: application.id, source } });
    const byKey = new Map(existing.map((route) => [route.routeKey, route]));

    for (const definition of request.routes) {
      const key = definition.key.trim();
      const path = definition.path.trim();
      const titles = Object.fromEntries(
        Object.entries(definition.titles || {})
          .map(([locale, title]) => [locale.trim(), title.trim()])
          .filter(([locale, title]) => Boolean(locale && title)),
      );
      if (!key || keys.has(key)) {
        throw new BadRequestException(`Route keys must be unique and non-empty: ${key || '(empty)'}.`);
      }
      if (!path.startsWith('/')) {
        throw new BadRequestException(`Route path must start with "/": ${key}.`);
      }
      if (Object.keys(titles).length === 0) {
        throw new BadRequestException(`At least one localized title is required: ${key}.`);
      }
      keys.add(key);

      const route = byKey.get(key) ?? this.routeRepo.create({
        id: uuidv4(),
        applicationId: application.id,
        source,
        routeKey: key,
      });
      route.pathTemplate = path;
      route.titles = titles;
      route.status = TenantRouteStatus.AVAILABLE;
      route.icon = definition.icon?.trim() || null;
      route.cssClass = definition.cssClass?.trim() || null;
      route.metadata = definition.metadata ?? null;
      route.lastSyncedAt = now;
      await this.routeRepo.save(route);
    }

    let unavailable = 0;
    if (request.replaceMissing !== false) {
      for (const route of existing.filter((entry) => !keys.has(entry.routeKey))) {
        route.status = TenantRouteStatus.UNAVAILABLE;
        route.lastSyncedAt = now;
        await this.routeRepo.save(route);
        unavailable += 1;
      }
    }

    return {
      applicationId: application.id,
      source,
      synchronized: keys.size,
      unavailable,
      synchronizedAt: now.toISOString(),
    };
  }
}
